### PR TITLE
Add an option to provide the sparsity pattern of jacobians and hessians

### PR DESCRIPTION
**Pull request by amontoison**
@jbcaillau

### CODE REVIEW

**Comment by jbcaillau** [comment]
> @jbcaillau

Thanks @amontoison So, provide: 
- sparse AD Jacobian (forward)
- sparse AD Hessian (forward and reverse) ?

**Comment by amontoison** [comment]
> > @jbcaillau
> 
> Thanks @amontoison So, provide:
> 
>     * sparse AD Jacobian (forward)
> 
>     * sparse AD Hessian (forward and reverse) ?

@jbcaillau
The sparsity pattern of Jacobians and Hessians is not related to AD but we must store it for the decompression step.
The AD is just used to recover the value of the nnz coefficients and we can use what we want for the backend.
The supported sparse backend are:
- `SparseADJacobian` -- Forward (Jacobian);
- `SparseADHessian` -- Forward over forward (Hessian);
- `SparseReverseADHessian` -- Forward over Reverse (Hessian).

In the future I would like to only support two backends (`SparseADJacobian` and `SparseADHessian`) with options for the "mode" (forward, reverse, both, forward over reverse, ...) and the "AD package" (Enzyme.jl, ForwardDiff.jl, ReverseDiff.jl, ...).

**Comment by jbcaillau** [comment]
> @jbcaillau The sparsity pattern of Jacobians and Hessians is not related to AD but we must store it for the decompression step. The AD is just used to recover the value of the nnz coefficients and we can use what we want for the backend [...]

@amontoison OK. Thought it was more directly related to this request (example / tuto on how to build an NLPModel)
https://github.com/control-toolbox/CTDirect.jl/issues/183

File: src/sparse_jacobian.jl [hunk 1]
```diff
@@ -22,7 +22,19 @@ function SparseADJacobian(
 ) where {T}
```
Comment (tmigot) [inline src/sparse_jacobian.jl:22]:
```suggestion
)
```
and adapt line 18

File: src/sparse_hessian.jl [hunk 2]
```diff
@@ -25,8 +25,21 @@ function SparseADHessian(
   detector::AbstractSparsityDetector = TracerSparsityDetector(),
   kwargs...,
 ) where {S}
```
Comment (tmigot) [inline src/sparse_hessian.jl:27]:
```suggestion
)
```
and adapt line 23

File: src/sparse_hessian.jl [hunk 3]
```diff
@@ -117,6 +130,19 @@ function SparseReverseADHessian(
   kwargs...,
 ) where {T}
```
Comment (tmigot) [inline src/sparse_hessian.jl:131]:
```suggestion
)
```
and adapt line 127v

**Review by tmigot — APPROVED** [review]
Thanks @amontoison ! Just a small improvement, I don't think we need the parametric type anymore.

**Comment by tmigot** [comment]
After, this PR I can rehabilitate this PR https://github.com/JuliaSmoothOptimizers/ADNLPModels.jl/pull/209 to have such tutorials

**Comment by amontoison** [comment]
@tmigot Do you think that we can add an option when we create an `ADNLPModel` / `ADNLSModel` to provide a sparsity pattern?
An option like `pattern_hessian` / `sp_hessian`.

**Comment by tmigot** [comment]
kwargs given to the ADNLPModels constructor are passed to the ADModelBackend which are passed to each backend.
Would there be a way to modify this PR so that it works?
Maybe add the sparsity pattern as kwargs with default of size (0,0) and then avoid calling the function only if size is different something like this?

I am not against adding kwargs, but ADNLPModels has so many constructors, I would love not adding more code if possible.

**Comment by amontoison** [comment]
Ok, let's not add new keywords and just add an example / tutorial in the documentation to provide the sparsity patterns.

**Pull request merged**